### PR TITLE
Feat/custom-template

### PR DESCRIPTION
**Pull request by jax-max**
# Add Custom Template Support for Change Files

## Summary

This PR introduces custom template support for OpenSpec change files, enabling projects to customize the structure and content of generated change proposal files (proposal.md, tasks.md, design.md, spec.md). The feature is fully backward compatible and opt-in, ensuring no impact on existing projects.

## Problem Statement

Currently, OpenSpec generates change files using hardcoded default structures. While these defaults work well for most cases, they present challenges for different project needs: large-scale projects find the default templates too simplistic and require more detailed sections, while some teams find alternative tools like spec-kit overly complex. Additionally, organizations need to adapt templates to match their specific documentation standards, compliance requirements, and review processes. The current hardcoded structure prevents teams from customizing document sections to fit their project scale and organizational needs.

## Solution

This implementation adds a flexible template system that:

1. Allows projects to define custom templates for each change file type
2. Supports variable substitution for dynamic content
3. Validates templates to ensure compatibility with OpenSpec's archive functionality
4. Falls back gracefully to default templates when custom templates are not available

### Core Implementation

- **New Module**: `src/core/templates/change-templates.ts` (192 lines)
  - `ChangeTemplateManager` class for template loading and rendering
  - Variable substitution engine for `{{changeId}}`, `{{date}}`, and `{{capability}}`
  - Spec template validation to ensure required tags are present
  - Default template definitions for all change file types

### CLI Enhancement

- **Modified**: `src/cli/index.ts`
  - Added `--templates` option to `openspec init` command
  - Optional flag to generate default templates during project initialization

- **Modified**: `src/core/init.ts`
  - Integrated template generation into initialization workflow
  - Added logic to skip generation if templates directory already exists
  - Non-intrusive implementation that does not affect default behavior

### Documentation Updates

- **Modified**: `src/core/templates/agents-template.ts`
  - Added "Custom Templates" section explaining template usage
  - Documented variable replacement syntax
  - Clarified fallback behavior and validation requirements

- **Modified**: `src/core/templates/slash-command-templates.ts`
  - Updated proposal steps to include template checking instructions
  - Added detailed template processing guidelines for AI assistants
  - Enhanced reference section with template variable documentation

### Testing

- **New**: `test/core/templates/change-templates.test.ts` (445 lines)
  - Comprehensive test coverage for all template operations
  - Tests for variable substitution, validation, and fallback behavior
  - Edge case handling and error scenarios
  - Optimized test structure using `loadTemplate()` for all scenarios

- **Modified**: `test/core/init.test.ts`
  - Added tests for template generation during initialization
  - Verified skip behavior when templates already exist

### Template System Architecture

The template system follows a layered approach:

1. **Template Discovery**: Checks for custom templates in `openspec/templates/` directory
2. **Template Loading**: Loads custom template or falls back to default
3. **Validation**: Validates spec templates to ensure archive compatibility
4. **Rendering**: Replaces variables with actual values
5. **Output**: Returns rendered template content

### Variable Substitution

Supported variables:
- `{{changeId}}`: The change identifier (e.g., `add-user-auth`)
- `{{date}}`: Current date in ISO format (YYYY-MM-DD)
- `{{capability}}`: Capability name (only used in spec.md templates)

### Spec Template Validation

Custom spec templates must include:
- At least one delta section header: `## ADDED|MODIFIED|REMOVED|RENAMED Requirements`
- Requirement header: `### Requirement:`
- Scenario header: `#### Scenario:`

Invalid templates trigger a warning and fall back to the default template to prevent breaking the archive functionality.

### Generating Default Templates

```bash
openspec init --templates
```

This command creates the following template files in `openspec/templates/`:
- `proposal.md.template`
- `tasks.md.template`
- `design.md.template`
- `spec.md.template`

### Customizing Templates

Users can edit the generated templates to match their project's documentation standards. The templates support variable substitution and will be automatically used when creating new change files.

### Template Processing

When AI assistants create change files, they:
1. Check for custom templates in `openspec/templates/`
2. Use custom template if available, otherwise use default structure
3. Replace variables with actual values
4. Validate spec templates before use

## Backward Compatibility

This feature is fully backward compatible:

- Default behavior unchanged: existing projects continue to work without modification
- Opt-in feature: templates are only generated when `--templates` flag is used
- Graceful fallback: if templates don't exist, system uses default structures
- No breaking changes: all existing APIs and workflows remain functional

## Testing

- All existing tests pass (33 tests in change-templates.test.ts)
- 445 lines of new test coverage
- No linter errors
- Comprehensive edge case coverage

## Files Changed

```
Added:
- src/core/templates/change-templates.ts (192 lines)
- test/core/templates/change-templates.test.ts (445 lines)

Modified:
- src/cli/index.ts (+3/-1 lines)
- src/core/init.ts (+29/-0 lines)
- src/core/templates/agents-template.ts (+32/-0 lines)
- src/core/templates/index.ts (+1/-0 lines)
- src/core/templates/slash-command-templates.ts (+30/-6 lines)
- test/core/init.test.ts (+100/-0 lines)

Total: +822/-11 lines across 8 files
```

## Summary by CodeRabbit

* **New Features**
  * Added an init option (--templates) to optionally generate default change templates; generation is skipped when custom templates exist. Templates support proposal, tasks, design and spec with variable substitution (changeId, date, capability) and spec-template validation with fallback to defaults.

* **Documentation**
  * Updated workflow docs with examples and template-driven file-creation guidance.

* **Tests**
  * Added tests for generation, loading, validation, substitution, skipping, and overwrite behavior.

### CODE REVIEW

File: src/core/templates/agents-template.ts [hunk 1]
```diff
@@ -42,9 +42,10 @@ Skip proposal for:
 
 **Workflow**
 1. Review \`openspec/project.md\`, \`openspec list\`, and \`openspec list --specs\` to understand current context.
-2. Choose a unique verb-led \`change-id\` and scaffold \`proposal.md\`, \`tasks.md\`, optional \`design.md\`, and spec deltas under \`openspec/changes/<id>/\`.
-3. Draft spec deltas using \`## ADDED|MODIFIED|REMOVED Requirements\` with at least one \`#### Scenario:\` per requirement.
-4. Run \`openspec validate <id> --strict\` and resolve any issues before sharing the proposal.
+2. Check for custom templates: If \`openspec/templates/\` directory exists, use those templates when creating files (see Custom Templates section below).
```
Comment (TabishB) [inline src/core/templates/agents-template.ts:45]: Hey @jax-max I try and reduce the amount of decisions an LLM has to make and try and make decisions here programatic if possible. 

I have something in mind on how to acccomplish this by creating an `openspec scaffold` command that sets up default templates for an Agent to fill out.

Check it out here: https://github.com/Fission-AI/OpenSpec/tree/main/openspec/changes/add-scaffold-command

Comment (jax-max) [reply src/core/templates/agents-template.ts:45]: Hi @TabishB I've seen the changes related to the scaffold – the last update was two months ago, so I'm not sure if the plan is still moving forward.
Personally, I think the scaffold should be project-scoped rather than change-scoped. Therefore, it would be better to handle template initialization during the openspec init phase: first generate the spec, task, design, and proposal templates, then modify these templates to meet the project's specific requirements. After that, all subsequent changes should adhere to these pre-configured templates.
I’ve also thought about making decisions programmatic if possible, but I don’t want to introduce additional commands (like bash or PowerShell) the way spec-kit does.
Do you have any better suggestions? Feel free to share them – I’m open to discussing this further!

Comment (jax-max) [reply src/core/templates/agents-template.ts:45]: I’ve also considered the spec file customization risk – We need to restrict users from modifying tags in spec deltas.

Comment (jax-max) [reply src/core/templates/agents-template.ts:45]: @TabishB  By the way, I’ve already dropped you a message on Discord, PTAL , hh

Comment (TabishB) [reply src/core/templates/agents-template.ts:45]: Hey @jax-max, I think we're on the same page here. The scaffold should ideally be project-scoped not change scoped, I agree!

The `openspec scaffold` should generate whatever the project defined template is (Should default to the standard OpenSpec template structure if it has not been modified.)

We can store this somewhere in `openspec/tempalate/*` 

The job of the scaffold command will be to just make a copy of the template into whatever change directory. (So should take <change-id> as input).

We then need to update the instructions to ask the agent to "fill in the documents" after creation.

I'll try and see if I can create a sequence diagram for this. We just need to be a bit careful with this change and we could possibly degrade the performance/quality of spec generation for current users. 

If there's another way to test this out without affecting the core flow, I'd be keen to integrate that first.

I'll work on the scaffold tomorrow and you can let me know if it works the way you expect?

Comment (jax-max) [reply src/core/templates/agents-template.ts:45]: hi @TabishB , That's awesome! I'm really looking forward to the scaffold feature. Regarding "asking the agent to 'fill in the documents' after creation", we actually prefer to handle this step manually. For large-scale backend projects, the agent’s analysis hasn’t been as effective as expected. Anyway, I’m still super excited for the scaffold to go live! Should I close the current PR?

**Comment by jax-max** [comment]
Duplicate with the scaffold's function
